### PR TITLE
Functions module

### PR DESCRIPTION
**Pull request by SotiriaSte**
**What kind of change does this PR introduce?** 

- [ ] Bugfix
- [X] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [X] No

**The PR fulfills these requirements:**

- [X] It's submitted to the `main` branch
- [ ] When resolving a specific issue, it's referenced in the PR's description  (e.g. `fix #xxx`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature 

**Other information:**

### CODE REVIEW

File: modules/functions/src/metrics/index.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { MetricType } from '@conduitplatform/grpc-sdk';
+
+export default {
```
Comment (kkopanidis) [inline modules/functions/src/metrics/index.ts:3]: add another metric to measure function execution times for each function. The function name or id will be defined via labels on the metric

File: modules/functions/src/admin/index.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        },
+      },
+      new ConduitRouteReturnDefinition('ListFunctionsDefinitions', 'String'),
+      this.getFunctionsDefinitions.bind(this),
+    );
+    this.routingManager.route(
+      {
+        path: '/definitions/:functionName',
+        action: ConduitRouteActions.GET,
+        description: 'Get function Definitions for specific function',
```
Comment (kkopanidis) [inline modules/functions/src/admin/index.ts:104]: these are logs/executions not definitions

File: modules/functions/src/controllers/function.controller.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  ConduitRouteReturnDefinition,
+  GrpcServer,
+  RequestHandlers,
+  RoutingManager,
+} from '@conduitplatform/grpc-sdk';
+import { Functions } from '../models';
+import { createFunctionRoute } from './utils';
+
+export class FunctionController {
+  private crudRoutes: {
```
Comment (kkopanidis) [inline modules/functions/src/controllers/function.controller.ts:12]: we don't have crud and custom routes here.

File: modules/functions/src/controllers/function.controller.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      })
+      .catch((err: Error) => {
+        ConduitGrpcSdk.Logger.error(
+          'Something went wrong when loading functions to the router',
+        );
+        ConduitGrpcSdk.Logger.error(err);
+      });
+  }
+
+  addRoutes(
```
Comment (kkopanidis) [inline modules/functions/src/controllers/function.controller.ts:73]: this is not needed here, add the routes in the object in a proper manner.

File: modules/functions/src/controllers/utils.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  grpcSdk: ConduitGrpcSdk,
+) {
+  const logs: string[] | undefined = [];
+  const vm = new NodeVM({
+    console: 'redirect',
+    sandbox: {},
+    timeout: timeout,
+    require: {
+      external: true,
+      import: ['lodash'],
```
Comment (kkopanidis) [inline modules/functions/src/controllers/utils.ts:44]: make sure that lodash is in package.json, and also add here and in package.json "axios".

File: modules/functions/src/controllers/utils.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  let duration;
+  let start;
+  try {
+    const script = `module.exports = function(grpcSdk,req,res) { ${functionCode} }`;
+    start = process.hrtime();
+    const functionInSandbox = vm.run(script);
+    const end = process.hrtime(start);
+    duration = end[0] * 1e3 + end[1] / 1e6;
+    const functionData = functionInSandbox(grpcSdk, call.request, callback);
+    await addSchemaDefinitions(name, duration, true, undefined, logs);
```
Comment (kkopanidis) [inline modules/functions/src/controllers/utils.ts:60]: function executions are not schema definitions

File: modules/functions/src/controllers/utils.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+  let duration;
+  let start;
+  try {
+    const script = `module.exports = function(grpcSdk,req,res) { ${functionCode} }`;
+    start = process.hrtime();
+    const functionInSandbox = vm.run(script);
+    const end = process.hrtime(start);
+    duration = end[0] * 1e3 + end[1] / 1e6;
+    const functionData = functionInSandbox(grpcSdk, call.request, callback);
```
Comment (kkopanidis) [inline modules/functions/src/controllers/utils.ts:59]: the function  ends after line 59 has been executed, but you are measuring duration before.

**Review by kkopanidis — CHANGES_REQUESTED** [review]

File: modules/functions/src/admin/index.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    }
+    const timeoutValue = inputs.timeout ?? 180000;
+    const query = {
+      name,
+      functionCode,
+      inputs,
+      returns,
+      timeout: timeoutValue,
+    };
+    return Functions.getInstance().create(query);
```
Comment (kkopanidis) [inline modules/functions/src/admin/index.ts:151]: How is the new function going to be published? You only create the model, you do not refresh the routes

**Review by kkopanidis — CHANGES_REQUESTED** [review]

**Review by kkopanidis — APPROVED** [review]

**Review by ChrisPdgn — APPROVED** [review]

**Review by sdimitris — COMMENTED** [review]
It is my belief that the "/functions/upload" route's body should be evaluated due to the fact that if the input is incorrect, FunctionsExecutions will not be created.

Let me know if I'm wrong. @kon14  @ChrisPdgn @SotiriaSte  @kkopanidis

**Pull request merged**